projects/pgai/pgai/vectorizer/generate/function_parser.py

```diff
@@ -0,0 +1,204 @@
+            "regclass": "str",
+        }
+        base_type = type_mapping.get(self.type_name, "Any")
+        if self.is_array and not base_type.startswith("list"):
+            type_str = f"list[{base_type}]"
+        else:
+            type_str = base_type
+        if not self.is_required:
+            type_str = f"{type_str} | None"
+        return type_str
+
+
+@dataclass
+class PostgresFunction:
+    name: str
+    schema: str
+    parameters: list[PostgresParameter]
+    return_type: str
+
+
+def get_function_metadata(
+    conn: psycopg.Connection, function_names: list[str]
+) -> list[PostgresFunction]:
+    """Extract function metadata from Postgres catalog."""
+
+    query = """
+        SELECT
+            p.proname,
+            n.nspname,
+            p.proargnames,
+            p.pronargdefaults,
+            string_to_array(array_to_string(p.proargtypes, ' '), ' ') as argtypes,
+            p.proallargtypes,
+            p.proargmodes,
+            t.typname as return_type,
+            p.oid,
+            p.proargnames IS NULL as null_argnames,
+            p.proargtypes IS NULL as null_argtypes
+        FROM pg_proc p
+        JOIN pg_namespace n ON p.pronamespace = n.oid
+        JOIN pg_type t ON p.prorettype = t.oid
+        WHERE n.nspname = 'ai'
+          AND p.proname = ANY(%s)
+        ORDER BY p.proname;
+    """
+
+    functions: list[PostgresFunction] = []
+    with conn.cursor() as cur:
+        cur.execute(query, (function_names,))
+
+        results = cur.fetchall()
+        for row in results:
+            parameters: list[PostgresParameter] = []
+            param_names = row[2] or []  # type: ignore
+            count_of_default_params = row[3]
+            param_count = len(param_names)  # type: ignore
+            non_default_count = param_count - count_of_default_params
+            param_types = row[4] or []  # type: ignore
+
+            for i, name in enumerate(param_names):  # type: ignore
+                type_oid = int(param_types[i]) if i < len(param_types) else None  # type: ignore
+
+                # Get detailed type info
+                type_query = """
+                    SELECT t.typname,
+                           (CASE WHEN t.typtype = 'b' THEN t.typname
+                                WHEN t.typtype = 'a' THEN (SELECT at.typname
+                                                         FROM pg_type at
+                                                         WHERE at.oid = t.typelem)
+                                ELSE t.typname
+                            END) as base_type,
+                           t.typtype = 'a' as is_array
+                    FROM pg_type t
+                    WHERE t.oid = %s
+                """
+                cur.execute(type_query, (type_oid,))
+                type_info = cur.fetchone()
+
+                type_name = type_info[1]  # type: ignore
+                is_array = type_info[2]  # type: ignore
+
+                default = None
```

Comment:
Is this correct?

Reply:
This is actually a bit misleading thanks for questioning it. I removed the default value from the code base now.

The generated classes don't actually have the correct default values the to_sql function just laves any None value out so that the ai.xyz() call uses the sqls default value. So the `default` is always None in this script (otherwise I'd have to correctly parse the sql default values into its python representation, which I am saving on this way.